src/cache/src/index.ts

```diff
@@ -376,81 +377,74 @@ export class Cache extends LRUCache<
 
-    // We might not write the val file at all so we dont want to
-    // kick off this promise until we know we need to.
-    const writeVal = () =>
-      mkdirP.then(() => this.#writeFileAtomic(valFile, val))
+    // Helper to wait for the operations we always want to run (write the key file)
+    // combined with any other operations passed in.
+    const finish = async (ops: Promise<void>[]) => {
+      await Promise.all([writeKeyP, ...ops])
+    }
 
     if (!intFile) {
       // No integrity provided, just write the value and key files
-      await Promise.all([writeKeyP, writeVal()])
-      return
+      return finish([writeCacheFile(valFile, val)])
     }
 
-    // Handle different integrity scenarios
+    // Now we know that we have been passed an integrity value
+    // that we should attempt to use...somehow.
 
     const intStats = await success(stat(intFile))
     if (!intStats) {
       // Integrity file doesn't exist yet
-      // Write the value file and then link to a new integrity file
-      await Promise.all([
-        writeKeyP,
-        writeVal().then(() => link(valFile, intFile)),
+      // Write the value file and then link that to a new integrity file
+      return finish([
+        writeCacheFile(valFile, val).then(() =>
+          link(valFile, intFile),
+        ),
       ])
-      return
     }
 
-    // Check if the integrity file and value file are the same size
-    // This is a fast shortcut to check if the content is likely the same
-    const sameSize = intStats.size === val.length
-
     const valStats = await success(stat(valFile))
-    if (!valStats && sameSize) {
-      // Path doesn't exist but integrity is already what we want
-      // Try to link integrity file to val file
-      // If linking fails, continue on
-      if ((await success(link(intFile, valFile))) !== FAILURE) {
-        await writeKeyP
-        return
-      }
+    if (!valStats) {
+      // Value file doesn't exist but integrity does, we know they are the same
+      // because we trust the integrity file, so attempt to link and be done.
+      return finish([link(intFile, valFile)])
     }
 
-    const linked =
-      intStats.ino === valStats?.ino && intStats.dev === valStats.dev
-
-    if (linked && sameSize) {
-      // Integrity and val file are already linked and the same
-      // So nothing to do except write the key file
-      await writeKeyP
-      return
-    }
-
-    if (!linked && sameSize) {
-      // Integrity and val file are not linked but same size
-      // so its likely the same content so force link them
-      // because thats faster than writing the value file.
-      if (
-        (await success(this.#linkAtomic(intFile, valFile))) !==
-        FAILURE
-      ) {
-        await writeKeyP
-        return
-      }
+    // We now know that both the value and integrity files exist.
+
+    // TODO: we are handling different registries here which will likely
```

Comment:
I think this TODO is kind of unnecessary, tbh, and it might not be worth the hassle.

If the size matches, that means that it was the same kind of thing, likely from the same exact URL. So, "do nothing" is completely correct. And if it's not, ok, well, we have the link, so breaking the link is allowed.